### PR TITLE
Bootloader and Device OS use same HAL driver, Implement mirror button…

### PR DESCRIPTION
**Pull request by YutingYou**
### Problem
1. Bootloader uses Nordic driver for button and led while Device OS uses Particle HAL driver. In this PR, both bootloader and Deivce OS will use  Particle HAL driver.

2. Implement mirror button and mirror led.
#### Note:
Bootloader increases to 32808 Bytes

```
   text    data     bss     dec     hex filename
  32808     812    6856   40476    9e1c ../build/target/bootloader/platform-14-lto/bootloader.elf
```

#### 1. Mirror Button Test
1. Build and flash bootloader application 
2. Build and flash below example app
3. Press button or mirror button to enter listening mode and exit listening mode
4. Observe the console log, it outputs click counter and debounce time.
5. Enter safe mode and DFU mode in the bootloader.
6. Use mirror button D2 to repeat step 5

#### 2. Mirror LED Test
1. Build and flash bootloader application 
2. Build and flash below example app
3. Use three LED and connect them to A3, A4, A5 pin
4. Observe LED state in Device OS and bootloader

#### 3. Mirror LED unit test
Test application path: /user/tests/wiring/no_fixture/led.cpp
You should change A7 pin to A3, because A7 doesn't exit on Xenon.

### Example App

```c
 #include "application.h"

SYSTEM_MODE(MANUAL);
Serial1LogHandler   logHandler(115200, LOG_LEVEL_ALL);

static volatile int counter = 0;

void setup() {
    RGB.mirrorTo(A3, A4, A5, false, true);
    // RGB.mirrorDisable(true);

    System.buttonMirror(D2, FALLING, true);

    System.on(button_click, [](system_event_t ev, int clicks) -> void {
        counter++;
    });
}

void loop() {
    Log.info("counter: %d", counter);
    delay(1000);
}
```

### References

[CH20601]
[CH20613]
[CH19813]
[CH28529]
---

### Completeness

[Gen 3] Adds button and RGB LED mirroring support (#1590)

### CODE REVIEW

File: hal/inc/gpio_hal.h [hunk 1]
```diff
@@ -61,6 +61,7 @@ PinMode HAL_Get_Pin_Mode(pin_t pin);
 void HAL_GPIO_Write(pin_t pin, uint8_t value);
 int32_t HAL_GPIO_Read(pin_t pin);
 uint32_t HAL_Pulse_In(pin_t pin, uint16_t value);
+void HAL_GPIO_Toggle(uint16_t pin);
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply hal/inc/gpio_hal.h]: Fixed

File: wiring_globals/makefile [hunk 2]
```diff
@@ -7,6 +7,6 @@ TARGET_TYPE = a
 BUILD_PATH_EXT = $(BUILD_TARGET_PLATFORM)
 
 # the only true dependency is Wiring, the others are transitive dependencies
-DEPENDENCIES = wiring hal services system platform communication
+DEPENDENCIES = wiring hal services system platform communication dynalib
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply wiring_globals/makefile]: Fixed

File: platform/MCU/nRF52840/src/hw_config.c [hunk 3]
```diff
@@ -104,61 +104,29 @@ void Set_System(void)
...
-    SPARK_ASSERT(!hal_flash_init());
-    SPARK_ASSERT(!hal_exflash_init());
 }
 
 void Reset_System(void) {
     __DSB();
 
     SysTick_Disable();
 
-    sd_nvic_DisableIRQ(RTC1_IRQn);
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply platform/MCU/nRF52840/src/hw_config.c]: Fixed

File: hal/src/nRF52840/rgbled_hal.c [hunk 4]
```diff
@@ -2,191 +2,188 @@
...
-    nrf_pwm_sequence_set(NRF_PWM0, 0, &rgb_seq);
+static void set_led_value(Led_TypeDef led, uint16_t value) {
+    // mirror led configuration is changed, should reconfigure mirror LED
+    if (HAL_Leds[led].version != LED_CONFIG_STRUCT_VERSION) {
+        return;
+    }
 
-    nrf_pwm_enable(NRF_PWM0);
+    uint32_t pwm_max = Get_RGB_LED_Max_Value();
+    HAL_PWM_Write_Ext(HAL_Leds[led].pin, HAL_Leds[led].is_inverted ? (pwm_max - value) : value);
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply hal/src/nRF52840/rgbled_hal.c]: Fixed

File: hal/src/nRF52840/rgbled_hal_impl.h [hunk 5]
```diff
@@ -26,31 +26,16 @@
...
-            uint8_t is_hal_pin  : 1; // HAL or GPIO pin
-            uint8_t is_inverted : 1; // If LED is "inverted", active state is 0, instead of 1
-            uint8_t is_pwm      : 1; // PWM enabled LED
-        };
-        uint8_t flags;
-    };
-    uint8_t padding[14];
+    uint8_t             version;        // Struct version
+    uint16_t            pin;
+    uint8_t             is_inverted;    // If LED is "inverted", active state is 0, instead of 1
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply hal/src/nRF52840/rgbled_hal_impl.h]: Fixed

File: hal/src/nRF52840/button_hal.c [hunk 6]
```diff
@@ -189,28 +200,143 @@ uint8_t BUTTON_GetState(Button_TypeDef Button)
...
+        .debounce_time = 0,
+        .interrupt_mode = mode,
+    };
+
+    HAL_Buttons[BUTTON1_MIRROR] = conf;
+
+    BUTTON_Init(BUTTON1_MIRROR, BUTTON_MODE_EXTI);
+
+    if (pin == HAL_Buttons[BUTTON1].pin) {
+        LOG(WARN, "Pin %d shares the same EXTI as SETUP/MODE button", pin);
```
[the remark this answers is not in the document]
Comment (YutingYou) [reply hal/src/nRF52840/button_hal.c]: Good to know, removed.

File: services/inc/rgbled.h [hunk 7]
```diff
@@ -42,20 +43,6 @@ typedef enum
 #define LED_RGB           LED3_LED4_LED2
 #define LED_USER          LED1
 
-// FIXME: These should be pulled from platform_config.h
```
Comment (avtolstoy) [inline services/inc/rgbled.h]: The unit tests seem to be broken. Let's leave these `ifndef`s here with default values in addition to `platform_config.h` and create an empty `platform_config.h` for GCC platform in `platform/MCU/gcc/inc`.

Comment (YutingYou) [reply services/inc/rgbled.h]: Fixed

**Comment by avtolstoy** [comment]
https://github.com/particle-iot/device-os/pull/1590/commits/8b641107a3e44d32190bd7e853b4cd3f4cf68223 also fixes a 'red led stays slightly on' issue with common anode configuration (inverted = true).

**Comment by avtolstoy** [comment]
I've also noticed that we are always reinitializing PWM on every value change, which is bound to cause glitches. Why are we not using `nrfx_pwm_sequence_update()` / `nrfx_pwm_sequence_values_update()` etc?

**Comment by YutingYou** [comment]
Fixes PWM glitches caused by reinitialize PWM module when changing value. Reinitialization only happens when changing frequency or configuring a new PWM pin in a same PWM group.

before fix:
image

after fix:
image

This fix cannot pass unit test in `user/tests/wiring/no_fixture/pwm.cpp`, because of below limitation:
> Listed here are the main features of one PWM module:
>     Change of polarity, duty-cycle, and base frequency possibly on every PWM period

It means we have to wait for a period then the duty-cycle will update. `test(PWM_09_LowFrequencyAnalogWriteOnPinResultsInCorrectPulseWidth) ` will fail after testing 4-bit resolution and start to test 12-bit resolution, the error is big between these two resolution.

A dummy read of `pulseIn()` is needed after `analogWrite()`, should we add it to unit test?

**Comment by technobly** [comment]
@avtolstoy @YutingYou any updates on this PR?  I will add the changes noted for `no_fixture/pwm.cpp`

**Comment by technobly** [comment]
@avtolstoy @YutingYou any updates on this PR?  I will add the changes noted for `no_fixture/pwm.cpp` in a separate PR.

**Pull request merged**